File: backend/app/schemas/debug.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models import AuthType, DebugMode, HttpMethod


class FailureContext(BaseModel):
    method: HttpMethod
    url: str
    request_headers: dict = Field(default_factory=dict)
    request_body: str | None = None
    status_code: int | None = None
    response_headers: dict = Field(default_factory=dict)
    response_body: str | None = None
    latency_ms: float | None = None
    success: bool = False
    error: str | None = None


class DebugAnalyzeRequest(BaseModel):
    run_id: UUID | None = None
    context: FailureContext | None = None
    force: bool = False


class AgentTraceStep(BaseModel):
    agent: str
    output: str


class SingleDebugResult(BaseModel):
    cause: str
    fix: str
    confidence: str = "medium"


class OptimizedRequest(BaseModel):
    method: HttpMethod
    url: str
    headers: dict = Field(default_factory=dict)
    query_params: dict = Field(default_factory=dict)
    body: str | None = None
    auth_type: AuthType = AuthType.NONE
    auth_config: dict = Field(default_factory=dict)
    notes: str | None = None


class MultiAgentDebugResult(BaseModel):
    diagnosis: str
    root_cause: str
    suggested_fix: str
    validated_fix: str
    validation_passed: bool
    optimized_request: OptimizedRequest
    agent_trace: list[AgentTraceStep]


class DebugSessionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID | None
    mode: DebugMode
    cause: str | None
    fix: str | None
    diagnosis: str | None
    root_cause: str | None
    suggested_fix: str | None
    validated_fix: str | None
    optimized_request: dict
    agent_trace: list
    llm_used: bool
    created_at: datetime


class SingleDebugResponse(BaseModel):
    session_id: UUID
    llm_used: bool
    result: SingleDebugResult


class MultiAgentDebugResponse(BaseModel):
    session_id: UUID
    llm_used: bool
    result: MultiAgentDebugResult

File: frontend/src/types/debug.ts
export interface RequestRun {
  id: string;
  method: string;
  url: string;
  status_code: number | null;
  success: boolean;
  error: string | null;
  latency_ms: number;
  created_at: string;
}

export interface SingleDebugResult {
  cause: string;
  fix: string;
  confidence: string;
}

export interface AgentTraceStep {
  agent: string;
  output: string;
}

export interface OptimizedRequest {
  method: string;
  url: string;
  headers: Record<string, string>;
  query_params: Record<string, string>;
  body: string | null;
  auth_type: string;
  auth_config: Record<string, string>;
  notes: string | null;
}

export interface MultiAgentDebugResult {
  diagnosis: string;
  root_cause: string;
  suggested_fix: string;
  validated_fix: string;
  validation_passed: boolean;
  optimized_request: OptimizedRequest;
  agent_trace: AgentTraceStep[];
}

export interface SingleDebugResponse {
  session_id: string;
  llm_used: boolean;
  result: SingleDebugResult;
}

export interface MultiAgentDebugResponse {
  session_id: string;
  llm_used: boolean;
  result: MultiAgentDebugResult;
}

export interface DebugSession {
  id: string;
  run_id: string | null;
  mode: "single" | "multi_agent";
  cause: string | null;
  fix: string | null;
  diagnosis: string | null;
  root_cause: string | null;
  created_at: string;
  llm_used: boolean;
}
